### PR TITLE
Automated spring decompression and adding links to HK when starting MIP

### PR DESCRIPTION
**Pull request by henningonsbring**
This PR allows the start of a MIP analysis using `cg workflow mip-dna` independent of if fastq files are decompressed to spring or unlinked in housekeeper.

The following will happen when running the `cg workflow mip-dna` command in different situations:
- If all fastq files are decompressed and fastq files are linked -> MIP analysis will start
- If all fastq files are decompressed and fastq files are not linked -> fastq files will be linked and MIP analysis will start
- If fastq files are compressed and no decompression job for these files are running -> decompression will start and MIP analysis will not start
- If fastq files are compressed and decompression for these files are running -> decompression will not start and MIP analysis will not start

The main reason for this update is to support automated start of reruns which fastqs are often compressed. If the crontab just repeat `cg workflow mip-dna` over and over for a case, then eventually it will be ready to start.

Also in this PR @patrikgrenfeldt added a --retrive-flowcells flag, which allow the user to retrieve flowcells.

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh automated-spring`
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-clinical-api-stage.sh automated-spring`

### How to test
- [ ] Run `cg workflow mip-dna -c <case ID>` for a case which fastq files are decompressed and linked in housekeeper
- [ ] Run `cg workflow mip-dna -c <case ID>` for a case which fastq files are decompressed, but not linked in housekeeper
- [ ] Run `cg workflow mip-dna -c <case ID>` for a case which fastq files are compressed to spring
- [ ] Run `cg workflow mip-dna -c <case ID>` for a case where a spring decompression job already is running

### Expected test outcome
- [ ] MIP start when fastq files are decompressed and linked in housekeeper
- [ ] Take a screenshot and attach or copy/paste the output.

- [ ] MIP start and housekeeper links are added when fastq files are decompressed, but not linked in housekeeper
- [ ] Take a screenshot and attach or copy/paste the output.

- [ ] MIP do not start when fastq files are compressed to spring
- [ ] Spring decompression start when fastq files are compressed
- [ ] Take a screenshot and attach or copy/paste the output.

- [ ] MIP do not start when fastq files are compressed to spring and spring decompression is running for the case being started
- [ ] The spring decompression job that is running is not interrupted by starting MIP for the same case
- [ ] No additional spring decompression job starts when starting MIP for a case that is being decompressed
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch
- [ ] Inform to ...

### CODE REVIEW

**Comment by henningonsbring** [comment]
This PR is associated with: https://github.com/Clinical-Genomics/cgp-reruns/issues/1

File: cg/meta/workflow/mip.py [hunk 1]
```diff
@@ -485,3 +495,83 @@ def get_analyses_from_trailblazer(
...
+        hk_data_obj = self.hk.get_files(bundle=bundle, tags=[tags])
+        for obj in hk_data_obj:
+            hk_data_str.append(obj.path)
+        return hk_data_str
+
+    def collect_files_in_folder(self, folder: str) -> list:
+        files_in_folder = os.listdir(folder)
+        return files_in_folder
+
+    def get_other_format_in_same_folder(
```
Comment (moonso) [inline cg/meta/workflow/mip.py:510]: Don't forget to add docstrings that describes what the function does

File: cg/meta/workflow/mip.py [hunk 2]
```diff
@@ -485,3 +495,83 @@ def get_analyses_from_trailblazer(
...
+            query_basename = os.path.basename(query)
+            query_root = os.path.splitext(query_basename)[0]
+            files_in_query_folder = self.collect_files_in_folder(query_folder)
+            for file in files_in_query_folder:
+                if file.startswith(query_root) and file.endswith(target_format):
+                    redundant_target_paths.append(file)
+        target_paths = list(dict.fromkeys(redundant_target_paths))
+        return target_paths
+
+    def check_spring_files(self, family_obj: models.Family) -> bool:
```
Comment (moonso) [inline cg/meta/workflow/mip.py:526]: Some of these functionality might already exist in apps/crunchy-api or meta/compress-api

File: cg/meta/workflow/mip.py [hunk 3]
```diff
@@ -485,3 +495,83 @@ def get_analyses_from_trailblazer(
...
+        fastqs_of_interest = self.get_other_format_in_same_folder(
+            family_obj=family_obj, query_format="spring", target_format="fastq.gz"
+        )
+        fastqs_linked_in_hk = self.collect_hk_data(family_obj.internal_id, "fastq")
+        for fastq in fastqs_of_interest:
+            if not fastq in fastqs_linked_in_hk:
+                fastqs_to_link = True
+        return fastqs_to_link
+
+    def check_system_call(self, system_call: str) -> bytes:
```
Comment (moonso) [inline cg/meta/workflow/mip.py:558]: There is a module in cg utils/process for talking to the system

File: cg/meta/workflow/mip.py [hunk 4]
```diff
@@ -485,3 +495,83 @@ def get_analyses_from_trailblazer(
...
+        returned_from_system = run(system_call, capture_output=True).stdout
+        return returned_from_system
+
+    def check_spring_decompression_jobs(self, family_obj: models.Family) -> bool:
```
Comment (moonso) [inline cg/meta/workflow/mip.py:562]: You should ask this question to the compression api

File: cg/meta/workflow/mip.py [hunk 5]
```diff
@@ -485,3 +495,83 @@ def get_analyses_from_trailblazer(
...
+        is_spring_decompressing = False
+        for sample in family_obj.internal_id:
+            returned_from_system = self.check_system_call(
+                f"squeue -o %j | grep spring | grep {sample}"
+            )
+            if returned_from_system:
+                is_spring_decompressing = True
+        return is_spring_decompressing
+
+    def start_decompression(self, case_id: str) -> None:
```
Comment (moonso) [inline cg/meta/workflow/mip.py:572]: Should be done in compression api

File: cg/meta/workflow/mip.py [hunk 6]
```diff
@@ -485,3 +495,83 @@ def get_analyses_from_trailblazer(
...
+        os.system(f"cg decompress case {case_id}")
+
+    def link_fastq(self, case_id: str) -> None:
```
Comment (moonso) [inline cg/meta/workflow/mip.py:575]: Import and use functionality, don't run shell command

**Comment by henningonsbring** [comment]
This PR can be removed. Please check this PR instead: https://github.com/Clinical-Genomics/cg/pull/916

**Comment by henrikstranneheim** [comment]
NIce! Closing

**Pull request closed without merging**